### DISCUSSION TITLE
如何排除<section epub:type="bibliography">....</section> 之间的内容不翻译呢？

### DISCUSSION THREAD

**Discussion by GoTop** [bookfere/Ebook-Translator-Calibre-Plugin#261] [Q&A]
我想排除epub书籍中，`<section epub:type="bibliography">....</section> `之间的内容

当我在插件的“内容”-“忽略元素”里输入 section[epub:type="bibliography"]，保存时会提示：
`section [epub：type=“bibliography“］ 不是一个有效的 CSS 选择器。`

我问了一下gpt4，它说

> epub:type 是一个自定义属性，它不是 HTML 规范的一部分，所以在某些浏览器或 HTML 校验工具中可能会提示它不是一个有效的 CSS 选择器。然而，大多数现代浏览器都支持使用非标准的属性作为 CSS 选择器。
> 如果你遇到了问题，可以考虑使用 data-* 属性来代替非标准属性，或者确保你的文档类型允许这种自定义属性的使用。如果你确实需要使用 epub:type 并且在你的环境中有效，那么选择器 section[epub:type="bibliography"] 应该是可以工作的。

对于这种情况，应该使用什么样的css选择器来选择这样的代码呢？

或者我直接修改epub文件里的`<section epub:type="bibliography">....</section> `为`<section data-epub-type="bibliography">....</section> `，然后使用section[data-epub-type="bibliography"]来选择排除它？

**Comment by bookfere** [comment]
试试 `section[type="bibliography"]` 呢？里面的 `epub:` 属于 XML 命名空间，不属于 CSS 选择器。

**Comment by GoTop** [comment]
> 试试 `section[type="bibliography"]` 呢？里面的 `epub:` 属于 XML 命名空间，不属于 CSS 选择器。

在插件里里添加了section[type="bibliography"]，但还是不能过滤掉对应的内容

**Comment by GoTop** [comment]
将epub里bindex.xhtml里的<section epub:type="bibliography">改为 <section data-epub-type="bibliography">
![WX20240330-194438@2x](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/1784319/db025c4a-3ee3-479c-9201-842df7be8b32)


插件的内容-忽略元素 里 填写
```
section[data-epub-type="bibliography"]
body[data-epub-type="backmatter"]
```
![WX20240330-194349@2x](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/1784319/6618807b-eb16-4dec-82ae-604e41e02e52)

点击翻译后，还是有bindex.xhtml页面的内容出现：
![WX20240330-194525@2x](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/1784319/c9e4c3b0-30cd-49da-8c49-b4cd07f6753d)

epub文件在这里：
[111.epub.zip](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/files/14811788/111.epub.zip)

**Comment by bookfere** [comment]
原因是目前规则只应用于已抽取出来的元素。在抽取的过程中，由于这个 section 元素中没有内容，就被忽略并将其后代元素抽取出来了，因此规则没能生效。后面会针对这个用例优化一下抽取过程。

**Comment by GoTop** [comment]
> 原因是目前规则只应用于已抽取出来的元素。在抽取的过程中，由于这个 section 元素中没有内容，就被忽略并将其后代元素抽取出来了，因此规则没能生效。后面会针对这个用例优化一下抽取过程。


那对于我上面这个页面，如果需要设置忽略元素，需要设置css选择器为` li `还是`ol`?

**Reply by bookfere** [reply]
如果你想忽略整个章节，可以先选择某一项，通过右键菜单中的“选择所在章节”，即可通过“删除”忽略它所在的整个章节的内容。

**Reply by GoTop** [reply]
选择所在章节 这个功能好像没有什么反应

而且一般也不会整个章节都不翻译 只是不想翻译某些代码、参考目录、词汇表之类的部分

**Comment by Milor123** [comment]
@bookfere @GoTop 

如果我想让它不删除（或不替换）某些标签，该怎么办？

我在转换器上遇到了一个问题

我有一些类

span class=“font0”>
....
span class=“font17”>

它们是 17，但翻译插件将它们更改为 

calibre1
....
 h3 class=“calibre13”> 及其他

如果我添加某种例外

这种情况发生在我的一些书中，但这是完全荒谬的行为，因为它完全破坏了结构，标题可能最终变成 7 号字体、无法阅读的东西、非常大的字母，而我的电脑或手机的字体一般看起来非常小，这就像损坏的电子书一样。就像有些书在离开某些软件（如 “ABBYY”）后会出现无效格式。

![image](https://github.com/user-attachments/assets/27ba99a3-c783-49cb-a40c-a940eca9a16e)

上图是 “ABBYY ”电子书的一个例子，我试过更改名称等，但总是出错，唯一的办法是删除 CSS 文件，但会丢失大部分标题命名，因为翻译器插件会强制嵌入名为 “calibre1”、“calibre2 ”的新 CSS，并删除 HTML 代码中的默认 CSS。

所以我的问题是，如何告诉程序不要删除它们，或者我该怎么做？
